src/librustc_mir/build/matches/test.rs

```diff
@@ -148,11 +176,26 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                         -> Vec<BasicBlock> {
         let scope_id = self.innermost_scope_id();
         match test.kind {
-            TestKind::Switch { adt_def } => {
+            TestKind::Switch { adt_def, ref variants } => {
                 let num_enum_variants = self.hir.num_variants(adt_def);
-                let target_blocks: Vec<_> =
+                debug!("num_enum_variants: {}", num_enum_variants);
+                debug!("variants.len(): {}", variants.len());
+                debug!("variants: {:?}", variants);
+                let target_blocks: Vec<_> = if variants.into_iter().any(|b| {!b}) {
+                    let otherwise_block = self.cfg.start_new_block();
+                    debug!("basic block: {:?} is an otherwise block!", otherwise_block);
+                    (0..num_enum_variants).map(|i| 
+                        if variants[i] {
+                            self.cfg.start_new_block()
+                        } else {
+                            otherwise_block
+                        }
+                    )
+                    .collect()
+                } else {
                     (0..num_enum_variants).map(|_| self.cfg.start_new_block())
-                                          .collect();
+                                          .collect()
+                };
```

Comment:
I think I'd find something like this easier to parse:

```rust
let mut otherwise_block = None;
let mut otherwise = || { if otherwise_block.is_none() { otherwise_block = self.cfg.start_new_block()}; otherwise_block.unwrap() };
let target_blocks: Vec<_> = variants.iter().map(|&matched| {
    if matched {
        self.cfg.start_new_block()
    } else {
        otherwise()
    }
}).collect();
debug!("enum variants: {}, variants.len(): {}, otherwise_block: {:?}", num_enum_variants, variants.len(), otherwise_block);
```